### PR TITLE
Introduce reduce stages for pipelines

### PR DESCRIPTION
**Pull request by krishnangovindraj**
## Release notes: product changes
Introduce reduce stages for pipelines to enable grouped aggregates.

## Motivation
Introduce reduce stages for pipelines to enable grouped aggregates.

## Implementation
- Implement grouped reduce stages for pipelines.

### CODE REVIEW

File: executor/pipeline/reduce.rs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    while let Some(result) = iterator.next() {
+        grouped_reducer.accept(&result?, &context)?;
+    }
+    Ok(grouped_reducer.finalise())
+}
+
+struct GroupedReducer {
+    input_group_positions: Vec<VariablePosition>,
+    grouped_aggregates: HashMap<Vec<VariableValue<'static>>, Vec<ReducerImpl>>,
+    reused_group: Vec<VariableValue<'static>>,
```
Comment (flyingsilverfin) [inline executor/pipeline/reduce.rs:75]: is this `current_group`?

File: executor/pipeline/reduce.rs [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    sample_reducers: Vec<ReducerImpl>,
```
Comment (flyingsilverfin) [inline executor/pipeline/reduce.rs:76]: Maybe just `reducers`?

File: executor/pipeline/reduce.rs [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    let mut grouped_reducer = GroupedReducer::new(program);
+    while let Some(result) = iterator.next() {
+        grouped_reducer.accept(&result?, &context)?;
+    }
+    Ok(grouped_reducer.finalise())
+}
+
+struct GroupedReducer {
+    input_group_positions: Vec<VariablePosition>,
+    grouped_aggregates: HashMap<Vec<VariableValue<'static>>, Vec<ReducerImpl>>,
```
Comment (flyingsilverfin) [inline executor/pipeline/reduce.rs:74]: Please be careful with terminology... `aggregate` is 2.x terminology

File: executor/pipeline/reduce.rs [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        VariableValue::Value(value) => Some(value.clone().into_owned()),
+        VariableValue::Thing(Thing::Attribute(attribute)) => {
+            // As long as these are trivial, it's safe to unwrap
+            let snapshot: &Snapshot = &context.snapshot;
+            let value = attribute.get_value(snapshot, &context.thing_manager).unwrap();
+            Some(value.clone().into_owned())
+        }
+        _ => unreachable!(),
+    }
+}
```
Comment (flyingsilverfin) [inline executor/pipeline/reduce.rs:155]: Does this have to be an owned value you return? This will induce extra Clones you don't really need I think?

Comment (krishnangovindraj) [reply executor/pipeline/reduce.rs:155]: One is owned, the other isn't. I can wrap it in `MaybeOwns`. I can also likely make it return the primitive.

File: executor/pipeline/reduce.rs (hunk 3 above)
Comment (krishnangovindraj) [reply executor/pipeline/reduce.rs:74]: Fair. What's the word for the output of a reduction?

File: executor/pipeline/reduce.rs (hunk 2 above)
Comment (krishnangovindraj) [reply executor/pipeline/reduce.rs:76]: Unsure. I should probably re-create the reducers from the program instead but I figured a clone would be faster. This is an always empty set that can be cloned.

File: executor/pipeline/reduce.rs (hunk 1 above)
Comment (krishnangovindraj) [reply executor/pipeline/reduce.rs:75]: No, it's just a reused vec.

**Review by krishnangovindraj — COMMENTED** [review]
replies

**Review by krishnangovindraj — COMMENTED** [review]
Flush pending comments

File: executor/pipeline/reduce.rs (hunk 3 above)
Comment (flyingsilverfin) [reply executor/pipeline/reduce.rs]: reduction

File: executor/pipeline/reduce.rs (hunk 4 above)
Comment (flyingsilverfin) [reply executor/pipeline/reduce.rs]: Should be possible to make this not owned in any branch

File: executor/pipeline/reduce.rs (hunk 2 above)
Comment (flyingsilverfin) [reply executor/pipeline/reduce.rs]: `unitialised_reduce_executors`

File: ir/program/block.rs [hunk 5]
```diff
@@ -234,6 +237,26 @@ impl VariableRegistry {
             VariableOptionality::Optional => true,
         }
     }
+
+    // TODO: This is out of place
```
[the remark this answers is not in the document]
Comment (flyingsilverfin) [reply ir/program/block.rs]: delete TODO

**Pull request merged**